include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,429 @@
+
+        uint8_t* mem;
+        size_t currSz;
+        core::LinearAddressAllocator<uint32_t> addrAlctr;
+
+    public:
+        SBackingMemManager() : currSz(INITIAL_MEM_SIZE), addrAlctr(nullptr, 0u, 0u, ALIGNMENT, MAX_MEM_SIZE) {
+            mem = reinterpret_cast<uint8_t*>(_IRR_ALIGNED_MALLOC(currSz, ALIGNMENT));
+        }
+        ~SBackingMemManager() {
+            _IRR_ALIGNED_FREE(mem);
+        }
+
+        uint8_t* alloc(size_t bytes)
+        {
+            auto addr = addrAlctr.alloc_addr(bytes, ALIGNMENT);
+            //TODO reallocation will invalidate all pointers to nodes, so...
+            //1) never reallocate (just have reasonably big buffer for nodes)
+            //2) make some node_handle class that will work as pointer but is based on offset instead of actual address
+            if (addr+bytes > currSz) {
+                size_t newSz = currSz<<1;
+                if (newSz > MAX_MEM_SIZE) {
+                    addrAlctr.free_addr(addr, bytes);
+                    return nullptr;
+                }
+
+                void* newMem = _IRR_ALIGNED_MALLOC(newSz, ALIGNMENT);
+                memcpy(newMem, mem, currSz);
+                _IRR_ALIGNED_FREE(mem);
+                mem = reinterpret_cast<uint8_t*>(newMem);
+                currSz = newSz;
+            }
+
+            return mem+addr;
+        }
+    };
+
+protected:
+    ~IR()
+    {
+        //call destructors on all nodes
+        for (auto* root : roots)
+        {
+            core::stack<decltype(root)> s;
+            s.push(root);
+            while (!s.empty())
+            {
+                auto* n = s.top();
+                s.pop();
+                for (auto* c : n->children)
+                    s.push(c);
+
+                if (!n->deinited) {
+                    n->~INode();
+                    n->deinited = true;
+                }
+            }
+        }
+    }
+
+public:
+    IR() : memMgr() {}
+
+    void deinitTmpNodes()
+    {
+        for (INode* n : tmp)
+            n->~INode();
+        tmp.clear();
+    }
+
+    template <typename NodeType, typename ...Args>
+    NodeType* allocNode(Args&& ...args)
+    {
+        uint8_t* ptr = memMgr.alloc(sizeof(NodeType));
+        return new (ptr) NodeType(std::forward<Args>(args)...);
+    }
+    template <typename NodeType, typename ...Args>
+    NodeType* allocRootNode(Args&& ...args)
+    {
+        auto* root = allocNode<NodeType>(std::forward<Args>(args)...);
+        roots.push_back(root);
+        return root;
+    }
+    template <typename NodeType, typename ...Args>
+    NodeType* allocTmpNode(Args&& ...args)
+    {
+        auto* node = allocNode<NodeType>(std::forward<Args>(args)...);
+        tmp.push_back(node);
+        return node;
+    }
+
+    struct INode
+    {
+        enum E_SYMBOL
+        {
+            ES_MATERIAL,
+            ES_GEOM_MODIFIER,
+            ES_FRONT_SURFACE,
+            ES_BACK_SURFACE,
```

Comment:
you're breaking the AST if you allow this, so either:
A) Have separate roots for front and back surface
B) have each BSDF/Material node always contain an entry for front and back functions

Reply:
i think we'll have to go with B

Reply:
lets go with B then... exactly how much flexibility does Nvidia MDL allow with this Front and Backface stuff?

Reply:
actually B would be a back idea... because it introduces a binary choice at possibly every node in the AST

but the thing which chooses between back and front face is only a single variable, so if you were to "unroll" the AST for different cases you'd only end up with two ASTs.

Reply:
decided